packages/compiler/core/types.ts

```diff
@@ -325,15 +325,15 @@ export interface BlockComment extends TextRange {
 }
 
 export interface CadlScriptNode extends ContainerNode, BaseNode {
-  kind: SyntaxKind.CadlScript;
-  statements: Statement[];
-  file: SourceFile;
-  inScopeNamespaces: NamespaceStatementNode[]; // namespaces that declarations in this file belong to
-  namespaces: NamespaceStatementNode[]; // list of namespaces in this file (initialized during binding)
-  usings: UsingStatementNode[];
-  comments: Comment[];
-  parseDiagnostics: Diagnostic[];
-  printable: boolean; // If this ast tree can safely be printed/formatted.
+  readonly kind: SyntaxKind.CadlScript;
+  readonly statements: Statement[];
```

Comment:
```suggestion
  readonly statements: readonly Statement[];
```

Not going to comment on all of them, but the arrays need to be readonly too or otherwise you can still mutate them, just not assign a new array.